### DISCUSSION TITLE
Windows support?

### DISCUSSION THREAD

**Discussion by bozzio83** [henrygd/beszel#18] [General]
As the title, can the agent be installed on a windows pc without docker?

**Comment by henrygd** [comment]
I haven't released the exe files because I haven't checked that they work properly yet.

I can manually upload them to the latest release if someone wants to test. If they work then I'll include them in releases going forward.

They're available as Linux binaries, so the hub should work well enough in WSL, but the agent may not give you the right numbers. I'm not entirely sure how that works.

**Comment by bozzio83** [comment]
Okay thank you for this. I'm still learning a lot so I will let someone else test as I may not be much help in trouble shooting.  Also, is there a way to add more than one filesystem per agent? To monitor multiple disks?

**Comment by henrygd** [comment]
I added the Windows builds in case anyone wants to test:

https://github.com/henrygd/beszel/releases/tag/v0.1.0

Check #36 for status of multiple disks feature.

**Reply by srinthe702** [reply]
Don't see the Win builds- are they available for testing??

**Reply by henrygd** [reply]
Check the [latest releases](https://github.com/henrygd/beszel/releases) for a Windows agent. 

Please not that it's experimental and [may get flagged by AV](https://go.dev/doc/faq#virus). I'll try to improve the setup process and sort out the major issues over the next handful of minor releases. See [this article by Alex Kretzschmar](https://blog.ktz.me/using-beszel-to-monitor-windows/) for usage.

**Comment by bozzio83** [comment]
Okay, I'm not sure what I'm doing wrong but I don't see where to put in the public key.  I run the agent and a window pops up and quickly goes away and nothing happens.

**Comment by henrygd** [comment]
I think for the agent you'd need to run it from the terminal and do something like this:

```powershell
set KEY="..." && "C:\path\to\beszel-agent.exe"
```

I'm not a windows person though so maybe someone else can help. I'm not sure if it will work anyway so don't spend a lot of time on it please.

**Reply by utlilb** [reply]
Using the above will error with "ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain". It should just be a matter of finding the correct parameters to pass. I'll keep testing.

**Reply by lingwunb666** [reply]
> 使用上述方法将出现错误“ssh：握手失败：ssh：无法验证，尝试的方法 [无公钥]，没有剩余支持的方法”。这应该只是找到要传递的正确参数的问题。我会继续测试。

I successfully ran the agent on Windows Server 2008, but encountered the same problem as you, friend. Have you found a solution?

**Reply by lingwunb666** [reply]
> > 使用方法会出现错误“ssh：上述握手失败：ssh：无法验证，尝试的方法[无]，没有剩下支持的方法”。这应该只是找到要提交的正确参数的问题。我会继续测试。
> 
> 我在 Windows Server 2008 上成功运行了代理，但遇到了与您相同的问题，朋友。您找到解决方案了吗？

I found the reason, in set KEY=, it should not be with ‘’
Instead, it should be
set KEY=ssh-ed25519 xxxxxxxxxxxxxxxx

**Comment by ameer1234567890** [comment]
I can't seem to find the windows binary release. Help.!

**Reply by henrygd** [reply]
I removed it from the release because of #46 

Someone said it got flagged by Windows Defender. Which isn't surprising because it's running an SSH server and checking system information. Then they ran the exe through a website that checks against 75 security vendors and 1 of them said it found malware.

I ran known legitimate software through the site and got the same result, so it's meaningless, but at the same time I don't want other people to go through the same process.

It doesn't seem to be working correctly anyway so I'm going to hold off on distributing it to anyone until I have time to test it and fix the issues if they're fixable.

If you really want to try it anyway, you can clone the repository and compile the exe using the [instructions in the readme](https://github.com/henrygd/beszel/blob/main/readme.md#compiling).

```bash
GOOS=windows GOARCH=amd64 CGO_ENABLED=0 go build -ldflags "-w -s" .
```

**Comment by ameer1234567890** [comment]
I compiled and tried it on my Windows machine. Windows Defender immediately detected  it as a threat.

**Reply by henrygd** [reply]
Thanks, there may be no way around that in the end.

Apparently it works with WSL (see #76) so that's the best option for Windows as of now.

**Comment by ameer1234567890** [comment]
I am guessing it would give details of the guest OS when run in WSL. Am I mistaken?

**Comment by ameer1234567890** [comment]
I now build Windows binary myself and use it with exceptions added in Windows defender, It has been working quite well. There is just one feature that I miss. Windows hosts do not seem to report temperature. Is this intended?

**Reply by henrygd** [reply]
Thanks for letting me know! I still haven't done much testing on Windows.

It uses gopsutil to read sensors, which [does have Windows specific code to query sensors via WMI.](https://github.com/shirou/gopsutil/blob/master/sensors/sensors_windows.go)

Here's some code to test in isolation. If you try it, let me know what it outputs.

```go
package main

import (
	"context"
	"fmt"

	"github.com/shirou/gopsutil/v4/sensors"
)

func main() {
	temperatures, err := sensors.TemperaturesWithContext(context.Background())
	if err != nil {
		err.(*sensors.Warnings).Verbose = true
		panic(err)
	}
	for _, temp := range temperatures {
		fmt.Printf("%s: %.1f\n", temp.SensorKey, temp.Temperature)
	}
}
```

**Reply by ameer1234567890** [reply]
Output:
```
ACPI\ThermalZone\THM__0: 24.9
```

**Reply by ameer1234567890** [reply]
This is odd. Today I ran the `go run temp.go` and it gave me an error.

```
panic: interface conversion: error is *ole.OleError, not *common.Warnings

goroutine 1 [running]:
main.main()
        C:/Users/Ameer/Downloads/beszel/beszel/cmd/agent/temp.go:13 +0x153
exit status 2
```

**Reply by henrygd** [reply]
Interesting, looks like the error type is not the same on Linux and Windows.

Try removing the `err.(*sensors.Warnings).Verbose = true` line and rerun. It should give you the actual error if it still happens.

You can also run the agent in debug mode to see if it logs any lines with "Sensor error".

**Reply by ameer1234567890** [reply]
OK I finally caught the issue. Beszel needed to be run as administrator in Windows. Now it works perfectly, temperature data has started showing up in graph.

**Reply by henrygd** [reply]
Awesome, good to know. Thank you for figuring that out :+1:

**Reply by ameer1234567890** [reply]
Now, there is this one thing that I have been noticing for the past days. I have 2 Windows machines and both of them report a constant temperature reading ever since the beginning.

Machine 1: 56.85 °C
Machine 2: 24.85 °C

Screenshots:
![Screenshot 2024-11-01 004106](https://github.com/user-attachments/assets/816e7c6f-5273-4fc5-9de6-ad9a65d8231b)
![Screenshot 2024-11-01 004156](https://github.com/user-attachments/assets/9536164f-f21d-41b6-a3f5-cc624e8a9ae8)

PS: Machine 2 is my working laptop, so it does not have continuous data for longer.

**Reply by henrygd** [reply]
Maybe you can try running Open Hardware Monitor to see what sensors it can pick up.

Their docs include the below snippet, so maybe it can help Beszel find the sensor data too.

> The Open Hardware Monitor publishes all sensor data to WMI (Windows Management Instrumentation). This allows other applications to read and use the sensor information as well.

https://openhardwaremonitor.org/

**Reply by henrygd** [reply]
Apparently Open Hardware Monitor hasn't been updated in a while but was forked here:

https://github.com/LibreHardwareMonitor/LibreHardwareMonitor

**Reply by ameer1234567890** [reply]
LibreHardwareMonitor provides a more realistic set of values, none of which match the static value returned by beszel-agent.

![Screenshot 2024-11-01 014031](https://github.com/user-attachments/assets/b6a99dbe-bb7d-48a6-be1b-a3128de49e1c)
PS: This is from machine 2

**Reply by henrygd** [reply]
I don't know very much about Windows, but gopsutil uses WMI to get sensor data while Libre Hardware Monitor does not. WMI seems to be much less reliable depending on hardware.

Realistically, the chances of me being able to write a non-WMI implementation are close to zero. So it may just be a limitation with the Windows version.

According to [this reddit thread](https://www.reddit.com/r/PowerShell/comments/pjvoxm/get_cpu_temperature_wo_wmi/), LibreHardwareMonitor does export its sensor data to WMI at ROOT\LibreHardwareMonitor, but gopsutil seems to look only at "root/wmi" if I'm interpreting this correctly:

https://github.com/shirou/gopsutil/blob/master/sensors/sensors_windows.go#L25

So it may be possible to pull temps from LibreHardwareMonitor if gopsutil will accept a PR to allow customization of the query string. Then you could pass in "root/LibreHardwareMonitor" or whatever it needs to be. But you would still need to have LibreHardwareMonitor running to export the sensor data.

**Comment by nicedevil007** [comment]
Is it possible to add multiple drives in windows? I didn't figure out how this can work right now.

Here my Script that runs during the installation task of my PowerShellAppDeploymentToolkit:

```powershell
$NSSMPath = "nssm.exe"
$ServiceName = "beszelagent"
$beszelPath = "$envProgramFiles\beszel-agent"
$ExePath = "$envProgramFiles\beszel-agent\agent.exe"
$EnvVar = "KEY=ssh-ed25519 AAA##################"

if (-Not (Test-Path $beszelPath -PathType Container)) {
	Write-ADTLogEntry -Message  "Folder $beszelPath does not exist, creating it..."
	New-ADTFolder -Path "$beszelPath"
} else {
	Write-ADTLogEntry -Message "Folder $beszelPath already exists."
}

$Service = Get-Service -Name $ServiceName -ErrorAction SilentlyContinue

if ($Service) {
    Write-Host "The service '$ServiceName' exists. It will now be stopped and deleted..."

    # If the service is running, stop it
    if ($Service.Status -eq "Running") {
        Write-ADTLogEntry -Message "Stopping the service '$ServiceName'..."
        Start-ADTProcess -FilePath "$envSystem32Directory\sc.exe" -ArgumentList "stop $ServiceName"
        Start-Sleep -Seconds 3  # Wait time for stopping the service
    }

    # Delete service using sc.exe
    Write-ADTLogEntry -Message "Deleting the service '$ServiceName'..."
    Start-ADTProcess -FilePath "$envSystem32Directory\sc.exe" -ArgumentList "delete $ServiceName"
}

Write-ADTLogEntry -Message "Copying latest file '$ServiceName'..."
Copy-ADTFile -Path "$($adtSession.DirFiles)\agent.exe" -Destination "$envProgramFiles\beszel-agent"

Write-ADTLogEntry -Message "Setting path for the service '$ServiceName'..."
Start-ADTProcess -FilePath $NSSMPath -ArgumentList "install $ServiceName `"$ExePath`"" -UseShellExecute -Wait

Write-ADTLogEntry -Message "Setting SSH key for the service '$ServiceName'..."
Start-ADTProcess -FilePath $NSSMPath -ArgumentList "set $ServiceName AppEnvironmentExtra `"$EnvVar`"" -UseShellExecute -Wait

Write-ADTLogEntry -Message "Enabling autostart for the service '$ServiceName'..."
Start-ADTProcess -FilePath $NSSMPath -ArgumentList "set $ServiceName Start SERVICE_AUTO_START" -UseShellExecute -Wait

Write-ADTLogEntry -Message "Starting the service '$ServiceName'..."
Start-Service -Name "$ServiceName" -ErrorAction SilentlyContinue
```

**Reply by henrygd** [reply]
There's an issue about this here: #422

I'll try to look into it before the next release.

**Comment by Chinoman10** [comment]
How can I give the Win Beszel agent access to Docker stats?
Right now I'm running the Win-"native" agent to have more accurate host stats (_64 GB RAM instead of 32 as reported by the Docker agent, full network bandwidth reporting, same with CPU usage, Nvidia GPU stats, and so on_), but I'm also 'forced' to run the docker agent to get docker statistics on top of it as well. Running two agents for a single machine isn't that great, but it's not the end of the world, specially for such an early project 😅 but wanted to ask if there's a solution anyway 🙏

**Reply by henrygd** [reply]
I think you can use `daemon.json` to expose the API on the network, then set `DOCKER_HOST=tcp://127.0.0.1:2375` on the agent. You might need to use the WSL IP instead of `127.0.0.1`.

https://docs.docker.com/engine/daemon/remote-access/

Or use a socket proxy: https://github.com/Tecnativa/docker-socket-proxy

**Reply by v1rusnl** [reply]
I use [socat](https://github.com/simonrupf/docker-socat) with 2375 proxied to 23750 on the host because I have not the nerves to fiddle with the unreliable Docker Desktop settings in Windows. Pointing beszel env in nssm to `DOCKER_HOST=tcp://127.0.0.1:23750` works like charm! Thanks for the suggestion!

**Reply by v1rusnl** [reply]
@henrygd any chance for additional disks on the windows agent in the future?

I used [lsblk for Windows](https://github.com/tenox7/lsblk) to check out the mount points and tried to add EXTRA_FILESYSTEMS=PhysicalDrive0,PhysicalDrive1... to the beszelagent service envs, but it does not get recognized.

![image](https://github.com/user-attachments/assets/31b85cc4-41c4-4a84-b0ca-45e6ff87774c)

**Reply by henrygd** [reply]
Yes, hopefully very soon.

**Comment by henrygd** [comment]
There's a Windows install script available for testing now if anyone wants to take a look: https://github.com/henrygd/beszel/discussions/739

**Reply by Chinoman10** [reply]
Would be cool to be able to set the installation path tbh...
I have multiple disks and always install programs in the P: disk, as opposed to C: (_except for stupid programs which don't allow you to do it_).
So I had to manually check the installation script, see what it was doing and copy the commands myself, and I still had to cut-paste the agent from where WinGet installed it to (C:) and into the folder where I previously installed the agent (P:)... I assume this'll be a pain for next agent upgrades, unless I script something myself to semi-automate this myself.

**Comment by v1rusnl** [comment]
Windows Drives are working now. Well done @henrygd !

**Reply by henrygd** [reply]
Thanks, good to hear it's working!

**Reply by v1rusnl** [reply]
@henrygd One thing I recognized is I don't get any IO-Stats from the Windows drives (also not from root partition). Is this a known limitation?

**Reply by henrygd** [reply]
I only have one machine to test (laptop with Windows 11 Home) but disk I/O does populate for me.

Try setting `LOG_LEVEL=debug` and look for the line starting with `DEBUG Disk I/O` that gets logged on startup.

https://beszel.dev/guide/environment-variables#windows

**Reply by v1rusnl** [reply]
Here we go...

```
2025/05/01 14:28:39 DEBUG 0.11.1
2025/05/01 14:28:39 DEBUG Not monitoring ZFS ARC err="open /proc/spl/kstat/zfs/arcstats: The system cannot find the path specified."
2025/05/01 14:28:39 ERROR Error getting disk partitions err="\tError 0: Access is denied.\n"
2025/05/01 14:28:39 DEBUG Disk partitions="[{\"device\":\"C:\",\"mountpoint\":\"C:\",\"fstype\":\"NTFS\",\"opts\":[\"rw\",\"compress\"]} {\"device\":\"D:\",\"mountpoint\":\"D:\",\"fstype\":\"NTFS\",\"opts\":[\"rw\"]} {\"device\":\"E:\",\"mountpoint\":\"E:\",\"fstype\":\"NTFS\",\"opts\":[\"rw\",\"compress\"]} {\"device\":\"F:\",\"mountpoint\":\"F:\",\"fstype\":\"NTFS\",\"opts\":[\"rw\",\"compress\"]} {\"device\":\"T:\",\"mountpoint\":\"T:\",\"fstype\":\"NTFS\",\"opts\":[\"rw\",\"compress\"]} {\"device\":\"W:\",\"mountpoint\":\"W:\",\"fstype\":\"exFAT\",\"opts\":[\"rw\"]}]"
2025/05/01 14:28:39 ERROR Error getting diskstats err="Access is denied."
2025/05/01 14:28:39 DEBUG Disk I/O diskstats="map[C::{\"readCount\":1554163,\"mergedReadCount\":0,\"writeCount\":3961479,\"mergedWriteCount\":0,\"readBytes\":213270117376,\"writeBytes\":53291291136,\"readTime\":5120,\"writeTime\":2713,\"iopsInProgress\":0,\"ioTime\":0,\"weightedIO\":0,\"name\":\"C:\",\"serialNumber\":\"\",\"label\":\"\"} D::{\"readCount\":81896,\"mergedReadCount\":0,\"writeCount\":23839,\"mergedWriteCount\":0,\"readBytes\":19005924864,\"writeBytes\":21539905536,\"readTime\":423,\"writeTime\":94,\"iopsInProgress\":0,\"ioTime\":0,\"weightedIO\":0,\"name\":\"D:\",\"serialNumber\":\"\",\"label\":\"\"} E::{\"readCount\":38734,\"mergedReadCount\":0,\"writeCount\":17929,\"mergedWriteCount\":0,\"readBytes\":7017382400,\"writeBytes\":86675456,\"readTime\":76,\"writeTime\":17,\"iopsInProgress\":0,\"ioTime\":0,\"weightedIO\":0,\"name\":\"E:\",\"serialNumber\":\"\",\"label\":\"\"} F::{\"readCount\":15620,\"mergedReadCount\":0,\"writeCount\":15684,\"mergedWriteCount\":0,\"readBytes\":260541440,\"writeBytes\":6338838528,\"readTime\":22,\"writeTime\":59,\"iopsInProgress\":0,\"ioTime\":0,\"weightedIO\":0,\"name\":\"F:\",\"serialNumber\":\"\",\"label\":\"\"}]"
2025/05/01 14:28:39 INFO Root disk mountpoint=/ io=C:
2025/05/01 14:28:39 WARN Device not found in diskstats name=T:
2025/05/01 14:28:39 WARN Device not found in diskstats name=W:
```

**Reply by henrygd** [reply]
Interesting. I don't get the `Access is denied` error on my machine, so it might be a permissions thing.

I'll look further into it. You might want to try running as administrator as a test.

**Reply by v1rusnl** [reply]
Same error when running agent exe as Admin.

**Reply by Chinoman10** [reply]
![image](https://github.com/user-attachments/assets/2b4f4f18-1633-4ed2-9889-7a79a8c0d8a0)
![image](https://github.com/user-attachments/assets/fa91a7a5-53ed-423d-963b-39d013a0fe7d)

While the drives are working in terms of capacity, it doesn't appear to be reading the throughput well.
I ran some Steam game updates to test it out and got nothing on those charts (_and yes, I waited until the CPU & bandwidth spikes appeared caused by the highspeed download_). Not sure what can be the cause though.

**Reply by henrygd** [reply]
Thanks for the details. Added this to the roadmap and will figure it out after I clear a couple PRs that have been waiting a while.
